src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,349 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+#include "mococrw/symmetric_memory.h"
+
+#include <openssl/evp.h>
+
+#include "mococrw/error.h"
+#include "mococrw/symmetric_crypto.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw
+{
+
+using namespace openssl;
+
+class AESCipher::Impl
+{
+public:
+    Impl(SymmetricCipherMode mode, SymmetricCipherKeySize keySize, SymmetricCipherPadding padding,
+         const std::vector<uint8_t> &secretKey, const std::vector<uint8_t> &iv, Operation operation,
```

Comment:
This declaration forces a copy of secretKey and iv. Should we just make those non-references and `std::move()` them into the object to allow users not to make a copy?

Reply:
To make use of this optimization we need to change the public interface accordingly. We will avoid the copy of IV (16 bytes) but the interface would be inconsistent with other APIs we have which work with vectors of bytes.